tools/cldr-apps/src/org/unicode/cldr/web/CookieSession.java

```diff
@@ -55,68 +70,112 @@ public class CookieSession {
+     *
      * @return the number of milliseconds remaining before the user will be disconnected
      *         unless the user does something "active" before then.
      * 
      * Here, something "active" means anything that causes userDidAction() to be called.
+     *
+     * Called locally and by SurveyAjax.processRequest
      */
-    public long timeTillKick() {
+    public long millisTillKick() {
         if (!KICK_IF_INACTIVE) {
             return 1000000; // anything more than one minute, to prevent browser console countdown
         }
-        final long now = System.currentTimeMillis();
+        final long nowMillisSinceEpoch = System.currentTimeMillis();
         final boolean guest = (user == null);
-        long myTimeout; // timeout in seconds.
+        long myTimeoutSecs; // timeout in seconds.
 
-        if (user != null && UserRegistry.userIsTC(user)) {
-            myTimeout = 60 * 20; // 20 minutes
-        } else if (guest) {
+        if (guest) {
+            myTimeoutSecs = Params.CLDR_GUEST_TIMEOUT_SECS.value();
+            /*
+             * Allow twice as much time, if there aren't too many guests.
+             */
             if (!tooManyGuests()) {
-                myTimeout = 60 * 10; // 10 minutes
-            } else {
-                myTimeout = Params.CLDR_GUEST_INACTIVITY.value();
+                myTimeoutSecs *= 2;
             }
         } else {
+            myTimeoutSecs = Params.CLDR_USER_TIMEOUT_SECS.value();
+            /*
+             * Allow twice as much time, if there aren't too many users.
+             */
             if (!tooManyUsers()) {
-                myTimeout = 60 * 20; // 20 min
-            } else {
-                myTimeout = Params.CLDR_USER_INACTIVITY.value();
+                myTimeoutSecs *= 2;
             }
         }
 
-        final long remain = (1000 * myTimeout) - (now - lastAction);
+        final long remainMillis = (1000 * myTimeoutSecs) - (nowMillisSinceEpoch - lastActionMillisSinceEpoch);
 
-        if (remain < 0) {
+        if (remainMillis < 0) {
             return 0;
         } else {
-            return remain;
+            return remainMillis;
         }
     }
 
     /**
-     * When did the user last touch this session?
+     * The time (in millis since 1970) when the user last touched this session.
+     *
+     * Set only by touch(); returned by getLastBrowserCallMillisSinceEpoch().
+     *
+     * Compare lastActionMillisSinceEpoch.
+     */
+    private long lastBrowserCallMillisSinceEpoch;
+
+    /**
+     * Get the time (in millis since 1970) when the user last touched this session.
+     *
+     * Compare getLastActionMillisSinceEpoch.
+     *
+     * Called by SurveyMain and AdminAjax.jsp.
      */
-    public long last;
+    public long getLastBrowserCallMillisSinceEpoch() {
+        return lastBrowserCallMillisSinceEpoch;
+    }
 
+    /**
+     * TODO: clarify who calls this and why; the usage of durationDiff with millisTillKick appears dubious
```

Comment:
It shows up in server logs:

`Removed stale session {CookieSession#dpxCWYURRO-Atoqe9oSI02hSRFo, user=null, millisTillKick=0 sec., millisSinceLastBrowserCall=607175, millisSinceLastUserAction=607175}`

Reply:
wow, this is from more than 3 years ago!

the reason I had to ask "who calls this", is that if I ask Eclipse or IntelliJ who calls this toString method it "finds" a zillion callers and then aborts -- so my preference would be to use a different name for the method than toString, and reserve toString for cases where an override is really necessary

okay, I see it ends up with the right answer, though currentTimeMillis() is called twice (since elapsedTime also calls it) and a time is calculated that's in the past rather than the future, but only the difference matters, so all's well that ends well!

the name durationDiff isn't very intuitive for me (difference between two durations?), and it's hard to understand why, given a duration to begin with, it would be necessary to call currentTimeMillis() even once, let alone twice...